### PR TITLE
feat(frontend/backend/shared): add possibility to complete remove completion of a course section tm-161

### PR DESCRIPTION
**Pull request by tarmiel**
изображение

### CODE REVIEW

File: apps/backend/src/modules/section-statuses/section-status.repository.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		userId,
+	}: SectionStatusGetAllRequestDto): Promise<SectionStatusEntity[]> {
+		const sectionStatusModels = await this.sectionStatusModel
+			.query()
+			.whereIn("course_section_id", function () {
+				void this.select("id")
+					.from(DatabaseTableName.COURSE_SECTIONS)
+					.where("course_id", courseId);
+			})
+			.andWhere("user_id", userId)
```
Comment (fshabanov) [inline apps/backend/src/modules/section-statuses/section-status.repository.ts:90]:
```suggestion
			.whereIn("courseSectionId", function () {
				void this.select("id")
					.from(DatabaseTableName.COURSE_SECTIONS)
					.where("courseId", courseId);
			})
			.andWhere("userId", userId)
```

File: apps/backend/src/modules/section-statuses/section-status.service.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	public async create(
+		payload: SectionStatusAddRequestDto,
+	): Promise<SectionStatusResponseDto> {
+		const sectionStatus = await this.sectionStatusRepository.create(
+			SectionStatusEntity.initializeNew({
+				courseSectionId: payload.courseSectionId,
+				status: payload.status,
+				userId: payload.userId,
```
Comment (fshabanov) [inline apps/backend/src/modules/section-statuses/section-status.service.ts:34]: Destructure properties of `payload`

File: apps/frontend/src/modules/section-statuses/slices/section-statuses.slice.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+			state.dataStatus = DataStatus.FULFILLED;
+		});
+		builder.addCase(updateStatus.fulfilled, (state, action) => {
+			const previousStatuses = state.sectionStatuses.filter(
+				(status) => status.id !== action.payload.id,
+			);
+			state.sectionStatuses = [...previousStatuses, action.payload];
+			state.dataStatus = DataStatus.FULFILLED;
+		});
+		builder.addMatcher(
```
Comment (fshabanov) [inline apps/frontend/src/modules/section-statuses/slices/section-statuses.slice.ts:36]: Use `addCase` for each separately

File: packages/shared/src/modules/course-sections/libs/types/course-section-with-status-relation.type.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import { type SectionStatusResponseDto } from "../../../section-statuses/section-statuses.js";
+import { type CourseSectionDto } from "./course-section-dto.type.js";
+
+type CourseSectionWithStatusRelation = CourseSectionDto & {
```
Comment (fshabanov) [inline packages/shared/src/modules/course-sections/libs/types/course-section-with-status-relation.type.ts:4]:
```suggestion
type CourseSectionWithStatusDto = CourseSectionDto & {
```

File: packages/shared/src/modules/section-statuses/libs/validation-schemas/section-status-create-body.validation-schema.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { z } from "zod";
+
+import { SectionStatus } from "../enums/enums.js";
+
+const sectionStatusCreateBody = z
+	.object({
+		courseSectionId: z.coerce.number(),
```
Comment (fshabanov) [inline packages/shared/src/modules/section-statuses/libs/validation-schemas/section-status-create-body.validation-schema.ts:7]:
```suggestion
		courseSectionId: z.number(),
```
If it's for body we don't need coerce

File: packages/shared/src/modules/section-statuses/libs/validation-schemas/validation-schemas.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+export { sectionStatusCreateBody } from "./section-status-create-body.validation-schema.js";
+export { sectionStatusUpdateBody } from "./section-status-update-body.validation-schema.js";
+export { sectionStatusUpdateQuery } from "./section-status-update-query.validation-schema.js";
+export { sectionStatusesGetAllQuery } from "./section-statuses-get-all-query.validation-schema.js";
```
Comment (fshabanov) [inline packages/shared/src/modules/section-statuses/libs/validation-schemas/validation-schemas.ts:4]:
```suggestion
export { sectionStatusesGetAllQuery } from "./section-status-get-all-query.validation-schema.js";
```
Don't forget about the validation object variable name

File: apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx [hunk 7]
```diff
@@ -1,22 +1,98 @@
...
+	});
+
+	const handleToggle = useCallback(
+		(section: CourseSectionWithStatusRelation) => {
+			if (!section.status) {
+				return void dispatch(
+					sectionStatusActions.create({
+						courseSectionId: section.id as number,
+						status: SectionStatus.COMPLETED,
+						userId: user.id,
```
Comment (fshabanov) [inline apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx:46]: We don't need userId for create/update/delete, Backend already knows a user who sends requests

File: apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx [hunk 8]
```diff
@@ -1,22 +1,98 @@
...
+					sectionStatusId: section.status.id,
+				}),
+			);
+		},
+		[dispatch, user.id],
+	);
+
+	useEffect(() => {
+		void dispatch(
+			sectionStatusActions.getAll({ courseId: courseId, userId: user.id }),
```
Comment (fshabanov) [inline apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx:68]: No need to send userId for the current user, only send it if we are viewing another user's course

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/backend/src/modules/section-statuses/section-status.repository.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+		}
+
+		return SectionStatusEntity.initialize({
+			courseSectionId: sectionStatusModel.courseSectionId,
+			createdAt: sectionStatusModel.createdAt,
+			id: sectionStatusModel.id,
+			status: sectionStatusModel.status,
+			updatedAt: sectionStatusModel.updatedAt,
+			userId: sectionStatusModel.userId,
+		});
```
Comment (Vfivew) [inline apps/backend/src/modules/section-statuses/section-status.repository.ts:61]: The same, but it's up to you to choose
```suggestion
      return sectionStatusModel
          ? SectionStatusEntity.initialize({
                  courseSectionId: sectionStatusModel.courseSectionId,
                  createdAt: sectionStatusModel.createdAt,
                  id: sectionStatusModel.id,
                  status: sectionStatusModel.status,
                  updatedAt: sectionStatusModel.updatedAt,
                  userId: sectionStatusModel.userId,
            })
      : null;
```

**Review by Vfivew — APPROVED** [review]

File: apps/frontend/src/libs/components/checkbox/styles.module.css [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+.input:checked:disabled {
+	background-color: var(--color-text-100);
+}
+
+.input:checked::before {
+	position: absolute;
+	right: 1px;
+	bottom: -1px;
+	font-size: 13px;
+	color: #ffffff;
```
Comment (ShadowOfTheSpace) [inline apps/frontend/src/libs/components/checkbox/styles.module.css:43]:
```suggestion
	color: var(--main-white);
```

File: apps/frontend/src/modules/section-statuses/section-statuses-api.ts [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+		return await response.json<SectionStatusGetAllResponseDto>();
+	}
+
+	public async updateStatus(
+		id: number,
+		payload: SectionStatusUpdateRequestDto,
+	): Promise<SectionStatusResponseDto> {
+		const response = await this.load(
+			this.getFullEndpoint(`${SectionStatusesApiPath.ROOT}${id}`, {}),
```
Comment (ShadowOfTheSpace) [inline apps/frontend/src/modules/section-statuses/section-statuses-api.ts:63]:
```suggestion
			this.getFullEndpoint(SectionStatusesApiPath.$ID, {id: String(id)}),
```

File: apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx (hunk 8 above)
Comment (tarmiel) [reply apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx:68]: At this moment, I don't think that we can distinguish our courses from those of other users.

**Review by Sytnikovakatya — APPROVED** [review]

File: apps/backend/src/modules/section-statuses/section-status.controller.ts [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+	 *        content:
+	 *          application/json:
+	 *            schema:
+	 *              type: object
+	 *              properties:
+	 *                  status:
+	 *                    type: string
+	 *                    enum: [completed, in-progress]
+	 *      responses:
+	 *        201:
```
Comment (YourTeaChan) [inline apps/backend/src/modules/section-statuses/section-status.controller.ts:243]:
```suggestion
	 *        200:
```

File: apps/backend/src/modules/section-statuses/libs/constants/relation-name.constant.ts [hunk 13]
```diff
@@ -0,0 +1,3 @@
+const RELATION_NAME = "courseSections";
```
Comment (what1s1ove) [inline apps/backend/src/modules/section-statuses/libs/constants/relation-name.constant.ts:1]: use an enum here. think in advance in such cases

File: apps/backend/src/modules/section-statuses/section-status.repository.ts (hunk 9 above)
Comment (what1s1ove) [reply apps/backend/src/modules/section-statuses/section-status.repository.ts:61]: can be both

File: packages/shared/src/modules/section-statuses/libs/validation-schemas/section-status-update-body.validation-schema.ts [hunk 14]
```diff
@@ -0,0 +1,11 @@
+import { z } from "zod";
+
+import { SectionStatus } from "../enums/enums.js";
+
+const sectionStatusUpdateBody = z
+	.object({
+		status: z.enum([SectionStatus.COMPLETED, SectionStatus.IN_PROGRESS]),
```
Comment (what1s1ove) [inline packages/shared/src/modules/section-statuses/libs/validation-schemas/section-status-update-body.validation-schema.ts:7]:
```suggestion
		status: z.enum(Object.values(SectionStatus)),
```

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: apps/backend/src/modules/section-statuses/libs/constants/relation-name.constant.ts (hunk 13 above)
Comment (tarmiel) [reply apps/backend/src/modules/section-statuses/libs/constants/relation-name.constant.ts:1]: I did this at first, but enum with one field was recommended to convert in constant several times

File: apps/backend/src/modules/section-statuses/section-status.service.ts [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+	public async create(
+		payload: SectionStatusAddRequestDto & { userId: number },
+	): Promise<SectionStatusResponseDto> {
+		const { courseSectionId, status, userId } = payload;
+
+		const sectionStatus = await this.sectionStatusRepository.create(
+			SectionStatusEntity.initializeNew({
+				courseSectionId: courseSectionId,
+				status: status,
+				userId: userId,
```
Comment (fshabanov) [inline apps/backend/src/modules/section-statuses/section-status.service.ts:36]:
```suggestion
				courseSectionId,
				status,
				userId,
```

File: apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx (hunk 8 above)
Comment (fshabanov) [reply apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx:67]: As discussed, we will have `/users/:userId/courses/:courseId`, make sure that the current user cannot change the section status for other user's course

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/course-description/course-description.tsx [hunk 16]
```diff
@@ -54,7 +57,11 @@ const CourseDescription: React.FC = () => {
 			<div className={styles["wrapper"]}>
 				<CourseDetails course={course} courseSections={courseSections} />
 				{hasCourseSections && (
-					<CourseSections courseSections={courseSections} />
+					<CourseSections
+						courseId={Number(course.id)}
```
Comment (fshabanov) [inline apps/frontend/src/pages/course-description/course-description.tsx:61]: Isn't `course.id` already a number?

File: apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx [hunk 17]
```diff
@@ -1,22 +1,104 @@
-import { type CourseSectionDto } from "~/modules/course-sections/course-sections.js";
+import { DataStatus } from "~/libs/enums/data-status.enum.js";
```
Comment (fshabanov) [inline apps/frontend/src/pages/course-description/libs/components/course-sections/course-sections.tsx:1]:
```suggestion
import { DataStatus } from "~/libs/enums/enums.js";
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/course-description/course-description.tsx (hunk 16 above)
Comment (what1s1ove) [reply apps/frontend/src/pages/course-description/course-description.tsx:61]: from `useParams` it's a string

File: packages/shared/src/modules/section-statuses/libs/validation-schemas/section-status-create-body.validation-schema.ts [hunk 18]
```diff
@@ -0,0 +1,13 @@
+import { z } from "zod";
+
+import { getZodEnum } from "../../../../libs/helpers/helpers.js";
+import { SectionStatus } from "../enums/enums.js";
+
+const sectionStatusCreateBody = z
+	.object({
+		courseSectionId: z.number(),
+		status: z.enum(getZodEnum(Object.values(SectionStatus))),
```
Comment (what1s1ove) [inline packages/shared/src/modules/section-statuses/libs/validation-schemas/section-status-create-body.validation-schema.ts:9]: `Object.values(SectionStatus)` can we just use a type case here ? 

i do not like this helper

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: apps/frontend/src/pages/course-description/course-description.tsx (hunk 16 above)
Comment (fshabanov) [reply apps/frontend/src/pages/course-description/course-description.tsx:61]: Yes, but it is `courseId` from params, `course.id` comes from the `course` that we have in redux store

File: packages/shared/src/modules/section-statuses/libs/validation-schemas/section-status-create-body.validation-schema.ts (hunk 18 above)
Comment (tarmiel) [reply packages/shared/src/modules/section-statuses/libs/validation-schemas/section-status-create-body.validation-schema.ts:9]: The problem here is that zod.enum requires an array with at least one element, but with object.values we can't guarantee that. Simple solution is to cast Object.values(SectionStatus) as [string, ...string[]], but we lose enum type(attempt to type cast properly looks messy). This issue almost everywhere is solved by using such a helper and I don't actually find cleaner solution

**Review by GEOFARL — APPROVED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**